libraries/eosiolib/contracts/eosio/key_value_singleton.hpp

```diff
@@ -0,0 +1,129 @@
+#include "../../core/eosio/datastream.hpp"
+#include "../../core/eosio/name.hpp"
+
+#include <eosio/to_key.hpp>
+
+namespace detail {
+   constexpr inline size_t max_stack_buffer_size = 512;
+}
+
+namespace eosio {
+   namespace internal_use_do_not_use {
+      extern "C" {
+         __attribute__((eosio_wasm_import))
+         int64_t kv_erase(uint64_t db, uint64_t contract, const char* key, uint32_t key_size);
+
+         __attribute__((eosio_wasm_import))
+         int64_t kv_set(uint64_t db, uint64_t contract, const char* key, uint32_t key_size, const char* value, uint32_t value_size);
+
+         __attribute__((eosio_wasm_import))
+         bool kv_get(uint64_t db, uint64_t contract, const char* key, uint32_t key_size, uint32_t& value_size);
+
+         __attribute__((eosio_wasm_import))
+         uint32_t kv_get_data(uint64_t db, uint32_t offset, char* data, uint32_t data_size);
+      }
+   }
+
+   template <typename T, eosio::name::raw SingletonName, eosio::name::raw DbName = "eosio.kvram"_n>
+   class kv_singleton {
```

Comment:
this isn't really a "kv_singleton". "kv" is an implementation detail, let's just call it. single<T>,

Reply:
We already have a `singleton` class that was implemented using multi-index, which is why I stuck `kv` on the front.

Reply:
I'm open to a different idea for a name that doesn't conflict with that though.